src/lib/support/BitFlags.h

```diff
@@ -27,122 +27,225 @@
 
 #include <stdint.h>
 
+#include <type_traits>
 #include <utility>
 
 namespace chip {
 
 /**
  * Stores bit flags in a type safe manner.
  *
+ * @tparam FlagsEnum is an `enum` or (preferably) `enum class` type.
  * @tparam StorageType is the underlying storage type (like uint16_t, uint32_t etc.)
- * @tparam FlagsEnum is the typesafe flags setting
+ *         and defaults to the underlying storage type of `FlagsEnum`.
  */
-template <typename StorageType, typename FlagsEnum>
+template <typename FlagsEnum, typename StorageType = typename std::underlying_type_t<FlagsEnum>>
 class BitFlags
 {
 public:
     static_assert(sizeof(StorageType) >= sizeof(FlagsEnum), "All flags should fit in the storage type");
+    using IntegerType = StorageType;
 
-    BitFlags() {}
-    BitFlags(const BitFlags &) = default;
+    BitFlags() : mValue(0) {}
+    BitFlags(const BitFlags & other) = default;
     BitFlags & operator=(const BitFlags &) = default;
 
-    explicit BitFlags(FlagsEnum v) { Set(v); }
+    explicit BitFlags(FlagsEnum value) : mValue(static_cast<IntegerType>(value)) {}
+    explicit BitFlags(IntegerType value) : mValue(value) {}
 
-    explicit BitFlags(StorageType value) { SetRaw(value); }
+    template <typename... Args>
+    BitFlags(FlagsEnum flag, Args &&... args) : mValue(Or(flag, std::forward<Args>(args)...))
+    {}
+
+    template <typename... Args>
+    BitFlags(const BitFlags<FlagsEnum> & flags, Args &&... args) : mValue(Or(flags, std::forward<Args>(args)...))
+    {}
 
-    BitFlags & Set(FlagsEnum v)
+    template <typename... Args>
+    BitFlags(IntegerType value, Args &&... args) : mValue(value | Or(std::forward<Args>(args)...))
+    {}
+
+    /**
+     * Set flag(s).
+     *
+     * @param other     Flag(s) to set. Any flags not set in @a other are unaffected.
+     */
+    BitFlags & Set(const BitFlags & other)
```

Comment:
Should we implement operator |= and | instead?

I find declaring BitFlags for single constants a bit odd:

```cpp
// why would we need this:
static const BitFlags<CertValidateFlags> sIgnoreNotBeforeFlag(CertValidateFlags::kIgnoreNotBefore);
static const BitFlags<CertValidateFlags> sIgnoreNotAfterFlag(CertValidateFlags::kIgnoreNotAfter);

// when we can just use 
// CertValidateFlags::kIgnoreNotBefore
// CertValidateFlags::kIgnoreNotAfter)

directly
```

Reply:
Personally I agree, but this PR already exploded beyond what I thought it would be when I pulled on the loose string, so I drew the line at making the flag enums type-safe using the existing notation.

The TestChipCert usage does seems a bit strange. I assume the idea was to have shorter names to keep `sValidationTestCases` somewhat readable.

Reply:
Also, `Clear` would remain a bit ugly — `flags &= ~BitFlags<FlagsType>(kFlagValue)` — unless we define a free `operator~` on `enum class`es.